updates to standardise execution env use pages - macos (#6944)
circleci/circleci-docs: jekyll/_cci2/using-macos.md

description: "Learn how to configure a your jobs to run in the macOS execution environment."
version:
- Cloud
---

The macOS execution environment is used for iOS and macOS development, allowing you to test, build, and deploy macOS and iOS applications on CircleCI. The macOS executor runs jobs in a macOS environment and provides access to iPhone, iPad, Apple Watch and Apple TV simulators.

You can use the macOS execution environment to run your [jobs]({{site.baseurl}}/2.0/jobs-steps/) in a macOS environment on a virtual machine (VM). You can access the macOS execution environment by using the `macos` executor and specifying an Xcode version:

```yaml
jobs:
  build:
    macos:
      xcode: 12.5.1

    steps:
      # Commands will execute in macOS container
      # with Xcode 12.5.1 installed
      - run: xcodebuild -version
```

## Supported Xcode versions
{: #supported-xcode-versions }

{% include snippets/xcode-versions.md %}

## Available resource classes

> {: available-resource-classes }

Author: ```suggestion
{: #available-resource-classes }
```

Class                 | vCPUs | RAM
----------------------|-------|-----
medium                | 4 @ 2.7 GHz     | 8GB
macos.x86.medium.gen2 | 4 @ 3.2 GHz     | 8GB
large                 | 8 @ 2.7 GHz     | 16GB
macos.x86.metal.gen1                 | 12 @ 3.2 GHz     | 32GB
{: class="table table-striped"}

**Note:** The `macos.x86.metal.gen1` resource requires a minimum 24-hour lease. See the [Dedicated Host for macOS]({{ site.baseurl }}/2.0/dedicated-hosts-macos) page to learn more about this resource class.

```yaml
jobs:
  build:
    macos:
      xcode: 12.5.1
    resource_class: large
```

## macOS VM Storage
{: #macos-vm-storage }

> The amount of available storage on our macOS VMs depends on the resource class and Xcode image being used. The size of the Xcode images varies based on which tools are pre-installed.

Author: ```suggestion
The amount of available storage on CircleCI macOS virtual machines depends on the resource class and Xcode image being used. The size of the Xcode images varies based on which tools are pre-installed. The table below indicated how much storage will be available with various Xcode/resource class combinations. Also note the exceptions to this noted below.
```

      - run: xcodebuild -version
```

## Supported Xcode versions
{: #supported-xcode-versions }

{% include snippets/xcode-versions.md %}

## Available resource classes
{: #available-resource-classes }

{% include snippets/macos-resource-table.md %}

```yaml
jobs:
  build:
    macos:
      xcode: 12.5.1
    resource_class: large
```

## macOS VM Storage
{: #macos-vm-storage }

> The amount of available storage on CircleCI macOS virtual machines depends on the resource class and Xcode image being used. The size of the Xcode images varies based on which tools are pre-installed. The table below indicated how much storage will be available with various Xcode/resource class combinations. Also note the exceptions to this noted below.

Author: ```suggestion
The amount of available storage on CircleCI macOS virtual machines depends on the resource class and Xcode image being used. The size of the Xcode images varies based on which tools are pre-installed. The table below indicates how much storage will be available with various Xcode/resource class combinations. Also note the exceptions to this noted below.
```

> The amount of available storage on CircleCI macOS virtual machines depends on the resource class and Xcode image being used. The size of the Xcode images varies based on which tools are pre-installed. The table below indicates how much storage will be available with various Xcode/resource class combinations. Also note the exceptions to this noted below.

Author: FYI I made some changes to this intro to the table, and to the phrasing of the note to try and clear up the exceptions

11.*          | medium, large         | 23GB
11.*          | macos.x86.medium.gen2 | 23GB
12.*          | medium, large         | 30GB
12.*          | macos.x86.medium.gen2 | 30GB
13.*          | medium, large         | 23GB
13.*          | macos.x86.medium.gen2 | 89GB
{: class="table table-striped"}

> **Note:** Xcode 12.0.1, 12.4.0 and 12.5.1 have a minimum 100GB of available storage.

Author: ```suggestion
**Note:** If you specify Xcode 12.0.1, 12.4.0 and 12.5.1 you have a minimum 100GB of available storage.
```

11.*          | medium, large         | 23GB
11.*          | macos.x86.medium.gen2 | 23GB
12.*          | medium, large         | 30GB
12.*          | macos.x86.medium.gen2 | 30GB
13.*          | medium, large         | 23GB
13.*          | macos.x86.medium.gen2 | 89GB
{: class="table table-striped"}

> **Note:** Xcode 12.0.1, 12.4.0 and 12.5.1 have a minimum 100GB of available storage.

Author: Good spot this is confusing! I will find out and clear this up

> **Note:** Xcode 12.0.1, 12.4.0 and 12.5.1 have a minimum 100GB of available storage.

Author: This table is showing the available storage on the VM once all the macOS gubbins is installed. The note is calling out the exceptions that for these Xcode versions there will be at least 100GB of storage available. I'll update the text, let me know if it's still unclear.

**Note:** Xcode 12.0.1, 12.4.0 and 12.5.1 have a minimum 100GB of available storage.

## Image update cycle for the macOS executor
{: #using-the-macos-executor }

> Each `macos` job is run a fresh virtual machine, running a specified version macOS. We build a new image each time a new stable, or beta, version of Xcode is released by Apple and aim to get these deployed as soon as possible. Generally, the contents of a particular build image will remain unchanged, except in very exceptional circumstances we might be forced to re-build a container for a specific reason. Our goal is to keep your execution environment stable, and to allow you to opt-in to newer containers by setting the `xcode` key in your `config.yml` file.

Author: Thanks for pointing this out!

Unlike our stable images (which are frozen and will not change), once a new beta image is released it will overwrite the previous beta image until a GM (stable) image is released, at which point the image is frozen and no longer updated. If you are requesting an image using an Xcode version that is currently in beta, please expect it to change when Apple releases a new Xcode beta with minimal notice. This can include breaking changes in Xcode/associated tooling which are beyond our control.

To read about our customer support policy regarding beta images, please check out the following [support center article](https://support.circleci.com/hc/en-us/articles/360046930351-What-is-CircleCI-s-Xcode-Beta-Image-Support-Policy-).

### Apple silicon support
{: #apple-silicon-support }

It is possible to build Apple Silicon/Universal binaries using Xcode `12.0.0` and higher as Apple provides both the Intel (`x86_64`) and Apple Silicon (`arm64`) toolchains in this release. Cross-compiling Apple Silicon binaries on Intel hosts has an additional overhead and as a result compilation times will be longer than native compilation for Intel.

> Running or testing Apple Silicon apps natively is currently not possible as CircleCI build hosts are Intel-based Macs. Binaries will need to be exported as [artifacts]({{site.baseurl}}/2.0/artifacts/) for testing apps locally. Alternatively, [CircleCI runner]({{site.baseurl}}/2.0/runner-overview/#supported) can also be used to run jobs natively on Apple Silicon.

Author: I have not checked but I will!

> Running or testing Apple Silicon apps natively is currently not possible as CircleCI build hosts are Intel-based Macs. Binaries will need to be exported as [artifacts]({{site.baseurl}}/2.0/artifacts/) for testing apps locally. Alternatively, [CircleCI runner]({{site.baseurl}}/2.0/runner-overview/#supported) can also be used to run jobs natively on Apple Silicon.

Author: Yes still correct, checked with the execution team

```yaml
version: 2.1
jobs:
  build-and-test:
    macos:
      xcode: 12.5.1
    environment:
      HOMEBREW_NO_AUTO_UPDATE: 1
    steps:
      - checkout
      - run: brew install wget
```

## Supported build and test tools
{: #supported-build-and-test-tools }

With the macOS executor on CircleCI, it is possible to customize your build as needed to satisfy almost any iOS build and test strategy.

### Common test tools
{: #common-test-tools }

The following common test tools are known to work well on CircleCI:

* [XCTest](https://developer.apple.com/library/ios/documentation/DeveloperTools/Conceptual/testing_with_xcode/chapters/01-introduction.html)
* [Kiwi](https://github.com/kiwi-bdd/Kiwi)
* [KIF](https://github.com/kif-framework/KIF)
* [Appium](http://appium.io/)

### React Native projects
{: #react-native-projects }

React Native projects can be built on CircleCI using `macos` and `docker` executor types. For an example of configuring a React Native project, please see [our demo React Native application](https://github.com/CircleCI-Public/circleci-demo-react-native)

### Creating a `config.yml` File
{: #creating-a-configyml-file }

> The most flexible way to customize your build is to modify the CircleCI configuration for your project in `.circleci/config.yml`. This allows you to run arbitrary bash commands as well as utilise built-in features such as workspaces and caching. See the [Configuring CircleCI]( {{ site.baseurl }}/2.0/configuration-reference/) documentation for a detailed description of the structure of the `config.yml` file.

Reviewer: ```suggestion
The most flexible way to customize your build is to modify the CircleCI configuration for your project in `.circleci/config.yml`. This allows you to run arbitrary bash commands as well as use built-in features such as workspaces and caching. See the [Configuring CircleCI]( {{ site.baseurl }}/2.0/configuration-reference/) documentation for a detailed description of the structure of the `.circleci/config.yml` file.
```
